### PR TITLE
Einat/go tls inspect new binary

### PR DESCRIPTION
**Pull request by einat-stern**
### What does this PR do?

Adds inspection of new binaries, using symbol searching and lookup tables.

### Motivation
Attaching go-tls ebpf probes to new binaries.

### Additional Notes
This is a refactor & aplitting of a pre-done work from this [PR](https://github.com/DataDog/datadog-agent/pull/10161) with some bugfixes.

### Possible Drawbacks / Trade-offs
No tests are currently added, it should be done soon.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/go/bininspect/newproc.go [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+	case GoArchARM64:
+		tlsg, ok := i.symbols["runtime"]
+		if !ok || tls == nil {
+			return 2 * uint64(i.elf.arch.PointerSize()), nil
+		}
+
+		return tlsg.Value + uint64(i.elf.arch.PointerSize()*2) + ((tls.Vaddr - uint64(i.elf.arch.PointerSize()*2)) & (tls.Align - 1)), nil
+
+	default:
+		return 0, fmt.Errorf("binary is for unsupported architecture")
```
Comment (guyarb) [inline pkg/network/go/bininspect/newproc.go:186]:
```suggestion
		return 0, errors.New("binary is for unsupported architecture")
```

File: pkg/network/go/bininspect/newproc.go [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+	// Additionally, on all architectures other than x86_64 and x86 (in all Go versions),
+	// the runtime.g pointer is stored on a register.
+	// On x86_64 pre-Go 1.17 and on x86 (in all Go versions),
+	// the runtime.g pointer is stored in the thread's thread-local-storage.
+	var runtimeGInRegister bool
+	if i.elf.arch == GoArchX86_64 {
+		runtimeGInRegister = abi == GoABIRegister
+	} else {
+		runtimeGInRegister = true
+	}
```
Comment (guyarb) [inline pkg/network/go/bininspect/newproc.go:223]:
```suggestion
	runtimeGInRegister := true
	if i.elf.arch == GoArchX86_64 {
		runtimeGInRegister = abi == GoABIRegister
	}
```

File: pkg/network/go/bininspect/newproc.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+// depending on the ABI and architecture:
+// 1. If it has a dedicated register, this function gives the register number
+// 2. Otherwise, this function finds the offset in TLS that the pointer exists at.
+//
+// See:
+// - https://go.googlesource.com/proposal/+/master/design/40724-register-calling.md#go_s-current-stack_based-abi
+// - https://go.googlesource.com/go/+/refs/heads/dev.regabi/src/cmd/compile/internal-abi.md#amd64-architecture
+// - https://github.com/golang/go/blob/61011de1af0bc6ab286c4722632719d3da2cf746/src/runtime/runtime2.go#L403
+// - https://github.com/golang/go/blob/61011de1af0bc6ab286c4722632719d3da2cf746/src/runtime/runtime2.go#L436
+func (i *newProcessBinaryInspector) getGoroutineIDMetadata(abi GoABI) (GoroutineIDMetadata, error) {
```
Comment (guyarb) [inline pkg/network/go/bininspect/newproc.go:206]: Consider add UTs to ensure the function works correctly over different architecture
if that's too complex, please add jira ticket to handle it later

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/go/bininspect/newproc.go [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+		return nil, err
+	}
+
+	abi, err := FindABI(goVersion, arch)
+	if err != nil {
+		return nil, err
+	}
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (einat-stern) [reply pkg/network/go/bininspect/newproc.go:47]: oops

File: pkg/network/go/bininspect/newproc.go [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+
+			locations, err := FindReturnLocations(i.elf.file, symbol, uint64(offset))
+			if err != nil {
+				return nil, fmt.Errorf("could not find return locations for function %q: %w", funcName, err)
+			}
+
+			returnLocations = locations
+		}
+
+		parameters, err := funcConfig.paramLookupFunction(i.goVersion, string(i.elf.arch))
```
[the remark this answers is not in the document]
Comment (einat-stern) [reply pkg/network/go/bininspect/newproc.go:122]: oops

File: pkg/network/go/bininspect/newproc.go (hunk 3 above)
Comment (einat-stern) [reply pkg/network/go/bininspect/newproc.go:206]: I'll add a ticket for now and we'll address the tests as a separate task for the inspection

File: pkg/network/go/bininspect/newproc.go [hunk 6]
```diff
@@ -120,6 +117,9 @@ func (i *newProcessBinaryInspector) findFunctions(functions map[string]FunctionC
 		}
 
 		parameters, err := funcConfig.paramLookupFunction(i.goVersion, string(i.elf.arch))
+		if err != nil {
+			return nil, fmt.Errorf("failed finding parameter of function %q: %w", funcName, err)
```
Comment (guyarb) [inline pkg/network/go/bininspect/newproc.go:121]:
```suggestion
			return nil, fmt.Errorf("failed finding parameters of function %q: %w", funcName, err)
```

File: pkg/network/go/bininspect/utils.go [hunk 7]
```diff
@@ -131,3 +135,51 @@ func FindReturnLocations(elfFile *elf.File, sym elf.Symbol, functionOffset uint6
...
+	var sectionsToSearchForSymbol []*elf.Section
+
+	for i := range f.Sections {
+		if f.Sections[i].Flags == elf.SHF_ALLOC+elf.SHF_EXECINSTR {
+			sectionsToSearchForSymbol = append(sectionsToSearchForSymbol, f.Sections[i])
+		}
+	}
+
+	if len(sectionsToSearchForSymbol) == 0 {
+		return 0, fmt.Errorf("symbol %s not found in file - no sections to search", symbol)
```
Comment (guyarb) [inline pkg/network/go/bininspect/utils.go:159]:
```suggestion
		return 0, fmt.Errorf("symbol %q not found in file - no sections to search", symbol)
```

**Review by guyarb — CHANGES_REQUESTED** [review]

**Review by guyarb — APPROVED** [review]
LGTM

**Review by hmahmood — APPROVED** [review]

**Pull request merged**